---
title: "Finally, functional class scoring enrichment analysis of Infinium DNA methylation data"
date: "`r Sys.Date()`"
output:
  rmdformats::robobook:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    fig_width: 5
    fig_height: 5
bibliography: gmea.bib
csl: plos-computational-biology.csl
---

<strong>
Mark Ziemann<sup>1,2*</sup>,
Mandhri Abeysooriya<sup>2,3*</sup>,
Anusuiya Bora<sup>2</sup>,
Severin Lamon<sup>3</sup>,
Yen Wong<sup>4</sup>,
Jeffrey Craig<sup>4</sup>
</strong>

1. Burnet Institute, Melbourne, Australia

2. Deakin University, School of Life and Environmental Sciences, Geelong, Australia.

3. Deakin University, School of Exercise and Nutrition, Burwood, Australia.

4. Deakin University, School of Medicine, Geelong, Australia.

(*) Correspondence: m.ziemann@deakin.edu.au

| Author | ORCID |
| --- | --- |
| Mark Ziemann | 0000-0002-7688-6974 |
| Anusuiya Bora | 0009-0006-2908-1352 |

Source: https://github.com/markziemann/gmea

<br>

## Abstract

* Infinium methylation beadarrays remain one of the most popular techniques for studying genomic
methylation, but tools for downstream pathways analysis remain limited.

* Functional class scoring (FCS) is a group of pathway enrichment techniques that involve the ranking of
genes and evaluation of their collective regulation in biological systems.

* Here we propose three different approaches to apply FCS to Infinium data, which vary upon when the
differential methylation analysis is conducted, at the level of probes, genes or gene sets.

* Using real EPIC array data, we demonstrate that aggregation of probe methylation to gene level before
differential methylation analysis yields a high degree of true positives and low number of contradictory
results.

* Using matched RNA-seq data we evaluate the relationship of methylation differences at promoters and
gene bodies with RNA expression at the level of pathways.

* We apply this approach to two further contexts, where existing methods failed to identify any
pathway-level effects in 450K and EPIC datasets: (i) the effect of assisted reproductive technology on
DNA methylation, and (2) the association of elevated plasma homocysteine with methylation patterns.

* The methods described here have been incorporated into a package called gmea, short for gene
methylation enrichment analysis.

**Keywords:** Pathway analysis, functional enrichment analysis, Infinium Array, DNA methylation,
epigenetics.

<br>

## Introduction

* DNA methylation is the most widely studied epigenetic mark, for its role in development and
disease.

* Many hundreds of epigenome-wide association studies (EWASs) are conducted each year to understand DNA
methylation patterns in disease and other contexts.

* Infinium arrays remain the preferred platform for EWASs due to low cost and analytical simplicity
as compared to high throughput methylation sequencing.

* Infinium arrays typically have multiple probes per gene in different locations including promoters,
CpG islands, gene bodies and enhancers, which complicates downstream functional enrichment analysis,
otherwise known as pathway analysis or gene set enrichment analysis.

* Enrichment analysis comes in two popular types; over-representation analysis (ORA) and functional
class scoring (FCS).

In ORA, genes with probes that meet an arbitrary significance threshold are selected and compared
to a background list consisting of genes measured in the assay.
The test seeks to identify sets of genes (eg: ontologies, pathways) that are over-represented in the
gene list of interest relative to the background.

Functional class scoring takes a different approach by ranking all detected genes by a differential
regulation score (eg: fold change, confident effect size) followed by a test to assess whether each
set of genes has distribution of scores that is different to the null.
Currently there are no specialised approaches for the application of FCS to Infinium array data.

* Here, we systematically evaluate different approaches to apply FCS to Infinium array.

* Due to the fact that genes have many probes, there are many concevable ways that FCS could be applied
to Infinium Methylation data.

* Method group 1: Limma is performed on the probe level and the probes beloning to each gene are scored
and these scores are aggregated to the gene level.
The gene scores undergo an enrichment test.

* Method group 2: The probe methylation values beloning to each gene are aggregated before differential
methylation analysis with limma.
Then, differential methylation results for each gene are scored, followed by an enrichment test.

With the above, there are options for how the aggregation steps are done, whether parametric or
non-parametric tests are used and whether the enrichment tests are self-contained or competitive.

To determine the best approach to applying FCS to Infinium methylation data, in this study we use
simulated data.
The proposed FCS methods are also compared against existing ORA methods.

## Methods

### Implementation overview

Functional enrichment analysis is a process of data summarisation from genes to gene sets (pathways).
This is made more complicated for Infinium array data due to the presence of multiple probes per gene,
meaning the data needs to be summarised from probes to genes and then to gene sets.
Here, we outline seven potential approaches for FCS of Infinium methylation data:

1. Limma Average t-test (LAT).
Differential methylation analysis is conducted at the probe level with limma, and the limma
t-statistics for each gene are summarised (mean).
The mean t-statistics are used in a downstream two-sample two-way t-test of gene set enrichment.

2. Limma Top t-test (LTT).
As above, except instead of calculating the mean t-statistic, the probe with the largest magnitude is selected
to represent the gene.

3. Limma Average Wilcox (LAW).
Similar to LAT, except instead of the two-sided t-test of gene set enrichment, a non-parametric alternative,
the Wilcoxon signed-rank test.

4. Limma Average Mitch (LAM).
Similar to LAW, however instead of the Wilcoxon test, the Mitch package is used to execute a two-way,
two-sample ANOVA-on-ranks test (Kruskal Wallis).

5. Aggregate Limma t-test (ALT).
In this method, all probe measurements beloning to each gene are averaged before conducting differential
methylation andlysis with limma.
The limma t-statistics are then used downstream for enrichment analysis with the two-sample two-way t-test.

6. Aggregate Limma Wilcox (ALW).
Similar to ALT, however it uses the Wilcoxon non-parametric test for gene set enrichment.

7. Aggregate Limma Mitch (ALM).
Similar to ALW, however uses the Mitch package to execute a two-way, two-sample ANOVA-on-ranks test
(Kruskal Wallis).

The above were compared to a standard ORA based approach (GSA), which involves limma on probes, selection of
statistically significant probes (FDR<0.05) for ORA with the `GSAmeth()` function of missMethyl which conducts a
modified hypergeometric test that accounts for multiple probe biases.
In the case that fewer than 250 significant probes are identified, then 250 probes with the smallest p-values
were selected.

### Method validation

To assess these methods, we adopted a simulated data approach.
EPIC Infinium array data derived from non-cancerous lung tissue of 37 cancer patients was used for our
evaluations.
We randomly sampled datasets to serve as control and cases (no replacement), with group sizes varying between
3 and 12.
One thousand random gene sets were created with sizes varying between 20 and 100 members.
Randow gene sets were used to avoid problems caused by the large overlap among real gene sets.
Throughout our evaluations, 50 gene sets were selected to be differentially methylated, 25 with increased and 25
with decreased methylation.
From these gene sets, half of the members were selected.
For those selected genes, half of the annotated probes were selected.
We adjusted the m-values in the case group by a specified amount, which we call the "delta", which we varied
between 0.1 and 0.4.
Following the incorporation of methylation changes to selected probes in the case group, the data underwent
enrichment analysis using `GSAmeth` approach, or one of the seven FCS methods outlined above.
Enrichment results were then filtered for FDR<0.05 and correct direction of methylation change.
Differentially methylated gene sets observed were compared to gene sets selected to be the ground truth, so that
the number of true positives (TP), false positives (FP) and false negatives (FN) could be determined.
At each setting for group size, delta and gene set size, 20 replications were conducted each with a set seed
to ensure reproducibility.
Mean TP, FP and FN values were determined, and then precision, recall and F1 scores were calculated.









As there are no established methods for benchmarking Infinium array based pathway regulation using
known ground truth synthetic data, we used a split-and-analyse approach with a real dataset to
evaluate the three approaches outlined above.
We downloaded raw intensity EPIC IDAT files from GEO for study GSE158422 which consists of
lung tumour and normal adjacent tissues from 37 patients.
Probe annotations were obtained from the "IlluminaHumanMethylationEPICanno.ilm10b4.hg19" Bioconductor
package.
Reactome gene sets were downloaded from the MSigDB website.
From this set of samples, 18 patients were randomly selected into analysis A, where control and tumour
samples were compared.
From the remaining 19 samples, 18 were randomly selected into analysis B, where control and tumour
samples were compared as above to act as an independent validation of analysis A.
Analyses A and B were conducted using one of the three enrichment approaches outlined above, and the
gene sets meeting a 5% FDR threshold in the hyper- and hypo-methylated directions were saved.
Gene sets identified to be common in A and B analyses (FDR<0.05 and direction) were considered validated,
gene sets identified in only one method were not considered validated, and gene sets that gave discordant
direction (hyper- vs hypo-methylation) between A and B were classed as contradictory.
This process was repeated 50 times to rigourously evaluate method performance.

### Sensitivity analysis

The same dataset including all 37 patients underwent analysis using the LA method, gsameth and ebGSEA.
The effect size (enrichment score) of significantly differentially methylated pathways (FDR<0.05) was
recorded.
Then, a random subset of n patients was selected followed by pathway enrichment analysis with the
respective method.
The subset results were compared to the full n=37 results to test the sensitivity of these enrichment
methods at differenent effect size levels.
This process was repeated 20 times.

### Association of methylation with gene expression

As methylation at promoters and gene bodies is considered to have differing association with gene
expression, we conducted LA analysis at gene promoters, gene bodies and whole gene separately and
examined the relationship with RNA-seq based pathway regulation.
RNA-seq counts were downloaded from the GEO database (Accession: GSE158420).
HGNChelper was used to repair transmogrified gene names, followed by DESeq2 based differential analysis.
Enrichment analysis of RNA-seq data was conducted using the mitch package with REACTOME gene sets.

### Pathway-level differential methylation in assisted reproductive technology
(i) the effect of assisted reproductive technology DNA methylation

### Pathway-level differential methylation in hyperhomocysteinemia
(2) the association of elevated plasma homocysteine with methylation patterns.

## Results

### Evaluation of FCS methods for Infinium methylation array data

Simulations were conducted using the seven FCS methods (LAT, LTT, LAW, LAM, ALT, ALW, ALM) and GSAmeth,
with group sizes of 3-12 and delta values between 0.1 and 0.5, and random gene sets with sizes 20, 50
and 100.
Precision and recall were calculated for these tests (Figure 1).
As expected, precision and recall values were better when group sizes and delta values were larger.


These data showed a strong relationship between the size of the gene sets and the precision and recall,
regardless of the enrichment test used.
On the whole, precision and recall were better 

We applied the three distinct approaches ("LA", "AL" and "AA") for pathway level differential
methylation analysis to a dataset consisting of lung tumour and adjacent normal tissue from 37 patients.
The dataset was split randomly into two groups of 18 patients and the overlap between subset was used to
assess the consistency and validity of findings with 50 simulations.
From the 1550 Reactome gene sets in the dataset, the AL method recorded the highest number of
"valid" pathways with an average of 919.8 being identified in both subsets, which was much higher than
LA and AA methods with 295.3 and 276.7 gene sets respectively (Figure 1A).
The AL method also recorded the lowest number of uncommon sets (382.0) which was better than the LA and
AA methods with 634.8 and 908.3 respectively (Figure 1B).
We also noticed that the LA method sometimes yielded contradictory results, so we quantified the number
of discordant observations, that is statistically significant but in the opposite direction.
True to this observation, the LA method gave 461.6 discordant pathways which indicates a high rate of
false-positives.
Meanwhile the AL and AA methods gave a mean of 0.1 and 0.02 discordant pathways respectively.
Taken together, these results indicate that the AL method gives the highest rate of common results
suggesting good accuracy in terms of sensitivity and specificity in this context.

### Comparision of AL method to existing pathway enrichment approaches


## Discussion

## Bibliography
