---
title: "GSE158422: DNA methylation in Lung squamous cell carcinoma - gene methylation enrichment analysis - sensitivity analysis"
author: "The GMEA team"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    fig_width: 7
    fig_height: 7
theme: cosmo
---

## Introduction

In the previous analysis, it was shown that the "AL" (aggregation-limma-enrichment) method
yielded good performance with a cohort of cancer-normal tissue pairs.

In this piece of work, we want to ascertain the sensitivity of this approach as compared to
existing methods such as gsameth and ebGSEA.

To achieve this we will analyse with the full 37 patient sample pairs, and repeatedly downsample
the dataset down to n=3 patients.
What we expect is that the "AL" method could have a higher degree of recall as compared to these
other approaches.
We expect that "AL" could identify gene sets with larger effect sizes (enrichment scores) using
fewer samples.
This means that for a given set of datasets, this approach could yield more findings, or
alternatively if pathway analysis is the primary outcome of a study, it need not be as large which
can save money and difficulty with establishing large sample collections.

## Requirements

This analysis was run on a 16C/32T computer with 128GB RAM running at 4.2 GHz.
RAM usage can be moderated by reducing the parallel core count.

```{r,packages}

suppressPackageStartupMessages({
  library("limma")
  library("eulerr")
  library("IlluminaHumanMethylation450kmanifest")
  library("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
  library("tictoc")
  library("mitch")
  library("kableExtra")
  library("beeswarm")
})

CORES=12

```

## Load data

* annotations

* probe sets

* gene sets

* design matrix

* mval matrix

```{r,annotation}

anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
myann <- data.frame(anno[,c("UCSC_RefGene_Name","Regulatory_Feature_Group","Islands_Name","Relation_to_Island")])

gp <- myann[,"UCSC_RefGene_Name",drop=FALSE]
gp2 <- strsplit(gp$UCSC_RefGene_Name,";")
names(gp2) <- rownames(gp)
sets <- split(rep(names(gp2), lengths(gp2)), unlist(gp2))

summary(unlist(lapply(sets,length)))

genesets <- gmt_import("https://ziemann-lab.net/public/gmea_prototype/ReactomePathways.gmt")

if (!file.exists("GSE158422_design.rds")) {
  download.file("https://ziemann-lab.net/public/gmea_prototype/GSE158422_design.rds", "GSE158422_design.rds")
}
design <- readRDS("GSE158422_design.rds")

if (!file.exists("GSE158422_design.rds")) {
 download.file("https://ziemann-lab.net/public/gmea_prototype/GSE158422_mx.rds","GSE158422_mx.rds")
}
mval <- readRDS("GSE158422_mx.rds")

boxplot(list("tumor"=matrix(colMeans(mval),ncol=2)[,1],"normal"=matrix(colMeans(mval),ncol=2)[,2]),
  main="mean probe methylation mval")

```

Reactome pathways were downloaded on the 14th Sept 2023 from MsigDB.

```{r,gsdl}

gs_entrez <- gmt_import("c2.cp.reactome.v2023.1.Hs.entrez.gmt")

gs_symbols <- gmt_import("c2.cp.reactome.v2023.1.Hs.symbols.gmt")

```

# AL GMEA: all probes

Start with curating the sample info.

```{r,samplesheet}

sex <- as.data.frame(design)$sex
tumor <- as.data.frame(design)$tumor
patient <- as.character(unlist(lapply(1:ncol(mval),function(i) {c(i,i)})))
patient <- head(patient,ncol(mval))
design <- model.matrix(~ patient + tumor )

```

## Functions

```{r,functions}

# magg is a general function that aggregates probe methylation measurements by gene
# in this variation it aggregates by MEAN, which performs better than median
magg <- function(mval,myann,cores=1){
  gn <- unique(unlist(strsplit( myann$UCSC_RefGene_Name ,";")))
  gnl <- strsplit( myann$UCSC_RefGene_Name ,";")
  gnl <- mclapply(gnl,unique,mc.cores=cores)
  myann$gnl <- gnl
  keep <- rownames(subset(myann,UCSC_RefGene_Name!=""))
  mx <- mval[rownames(mval) %in% keep,]
  mymed <- function(g) {
    probes <- rownames(myann[grep(g,myann$gnl),])
    rows <- which(rownames(mx) %in% probes)
    if ( length(rows) > 1 ) {
      b <- mx[rows,]
      med <- apply(b,2,mean)
      med <- matrix(med,nrow=1)
      colnames(med) <- colnames(b)
      rownames(med) <- g
      return(med)
    }
  }
  med <- mclapply(gn,mymed,mc.cores=cores)
  med <- med[lapply(med,length)>0]
  medf <- do.call(rbind,med)
  return(medf)
}

# chragg is a specialised function that performs the magg routine on chromosomes in parallel
chragg <- function(mval,myann,cores=1){
  annodf <- as.data.frame(anno)
  keep <- rownames(subset(myann,UCSC_RefGene_Name!=""))
  mx <- mval[rownames(mval) %in% keep,]
  chrs <- unique(anno$chr)
  myorder <- unlist(lapply(chrs,function(mychr) { nrow( annodf[annodf$chr==mychr,] ) } ))
  chrs <- chrs[order(-myorder)]
  leadercores <- floor(sqrt(cores))
  workercores <- ceiling(sqrt(cores))
  chrmedf <- mclapply(chrs,function(chr) {
    chrfrag <- annodf[annodf$chr==chr,]
    chrprobes <-rownames(chrfrag)
    chrmx <- mx[rownames(mx) %in% chrprobes,]
    chranno <- myann[rownames(myann) %in% chrprobes,]
    chrmedf <- magg(mval=chrmx,myann=chranno,cores=workercores)
    return(chrmedf)
  },mc.cores=leadercores)
  medf <- do.call(rbind, chrmedf)
  return(medf)
}

# agglimma is a helper function that runs limma analysis on the aggregated methylation data
# and fixes any duplicated gene names - most significant one is retained
agglimma <- function(medf,design) {
  fit.reduced <- lmFit(medf,design)
  fit.reduced <- eBayes(fit.reduced)
  dmagg <- topTable(fit.reduced,coef=ncol(design), number = Inf)
  nondup <- !duplicated(dmagg$ID)
  dmagg <- dmagg[nondup,]
  rownames(dmagg) <- dmagg$ID
  dmagg$ID = NULL
  return(dmagg)
}

# ttenrich is a function designed to conduct analysis of differential methylation.
# It takes the t-statistics from limma for a geneset and determines whether this gene
# set has a distribution different to the null (zero) using a one-sample two-sided t-test.
ttenrich <- function(m,genesets,cores=1) {
  res <- mclapply( 1:length(genesets), function(i) {
    gs <- genesets[i]
    name <- names(gs)
    n_members <- length(which(rownames(m) %in% gs[[1]]))
    if ( n_members > 4 ) {
      tstats <- m[which(rownames(m) %in% gs[[1]]),]
      myn <- length(tstats)
      mymean <- mean(tstats)
      mymedian <- median(tstats)
      wt <- t.test(tstats)
      res <- c(name,myn,mymean,mymedian,wt$p.value)
    }
  } , mc.cores = cores)
  res_df <- do.call(rbind, res)
  rownames(res_df) <- res_df[,1]
  res_df <- res_df[,-1]
  colnames(res_df) <- c("n_genes","t_mean","t_median","pval")
  tmp <- apply(res_df,2,as.numeric)
  rownames(tmp) <- rownames(res_df)
  res_df <- tmp
  res_df <- as.data.frame(res_df)
  res_df <- res_df[order(res_df$pval),]
  res_df$logp <- -log10(res_df$pval )
  res_df$fdr <- p.adjust(res_df$pval,method="fdr")
  res_df[order(abs(res_df$pval)),]
  return(res_df)
}

```

Run AL method on methylation data

```{}

# aggregation step
medf <- chragg(mval,myann,cores=CORES)

# differential analysis
dmagg <- agglimma(medf,design)
hist(dmagg$t)

head(dmagg, 30) %>%
  kbl(caption="limma gene methylation results") %>%
  kable_paper("hover", full_width = F)
# get t-stats
m <- as.data.frame(dmagg$t)
rownames(m1) <- rownames(magg1)
colnames(m1) <- "t"

# enrichment test
alres <- ttenrich(m=m,genesets=gs_symbols,cores=CORES)
hist(alres$t_mean)

alsig <- subset(alres,fdr<0.05)

nrow(alsig)

nrow(subset(alsig,t_median>0))

head(alsig[order(-alsig$t_median),],20) %>%
  kbl(caption="hypermethylated pathways") %>%
  kable_paper("hover", full_width = F)

nrow(subset(alsig,t_median<0))

head(alsig[order(alsig$t_median),],20) %>%
  kbl(caption="hypomethylated pathways") %>%
  kable_paper("hover", full_width = F)

```

## Overlap

Use Euler diagrams to find the similarity.
As medians are normally distributed about the zero, one would expect equal numbers of up and down-
methylated gene sets.
Therefore the t-test based approach could be considered better.

```{r,ol1}

lasig_up <- rownames(subset(lasig,t_median>0))
lasig_dn <- rownames(subset(lasig,t_median<0))

alsig_up <- rownames(subset(alsig,t_median>0))
alsig_dn <- rownames(subset(alsig,t_median<0))

v1 <- list("LA up"=lasig_up, "LA dn"=lasig_dn,
  "AL up"=alsig_up,"AL dn"=alsig_dn)

plot(euler(v1),quantities = TRUE)

```

# GMEA: promoter probes only

## Limma-aggregate-1-sample t-test promoter only

```{r,ls_promoter}

dmp <- dm[grep("Promo",dm$Regulatory_Feature_Group),]

tic()
dmpagg <- agg(dmp,cores=CORES)
toc()

head(dmpagg$gme_res_df,20) %>%
  kbl(caption="top gmea genes (t-test)") %>%
  kable_paper("hover", full_width = F)

```

## Session information

```{r,save}

save.image("GSE158422_gmea.Rdata")

sessionInfo()

```
