---
title: "GSE158422: DNA methylation in Lung squamous cell carcinoma - RNA expression and integration"
author: "The GMEA team"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    fig_width: 7
    fig_height: 7
theme: cosmo
---

## Introduction

Let's take a look at how pathway methylation associates with gene expression.

Partition methylation analysis between promoter and body.

Will need to do this at the annotation level.

Compare to GSAmeth.

```{r,packages}

suppressPackageStartupMessages({
  library("DESeq2")
  library("eulerr")
  library("IlluminaHumanMethylation450kmanifest")
  library("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
  library("HGNChelper")
  library("tictoc")
  library("limma")
  library("mitch")
  library("kableExtra")
  library("beeswarm")
  library("missMethyl")
  library("HGNChelper")
  library("gplots")
  library("gridExtra")
  library("png")
})

CORES=16

```

## Load annotations and methylation data

```{r,annotation}

anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
myann <- data.frame(anno[,c("UCSC_RefGene_Name","UCSC_RefGene_Group","Islands_Name","Relation_to_Island")])
gp <- myann[,"UCSC_RefGene_Name",drop=FALSE]
gp2 <- strsplit(gp$UCSC_RefGene_Name,";")
names(gp2) <- rownames(gp)
sets <- split(rep(names(gp2), lengths(gp2)), unlist(gp2))
summary(unlist(lapply(sets,length)))

design <- readRDS("GSE158422_design.rds")
mval <- readRDS("GSE158422_mx.rds")

runlimma <- function(mval,design,myann) {
  fit.reduced <- lmFit(mval,design)
  fit.reduced <- eBayes(fit.reduced)
  dm <- topTable(fit.reduced,coef=ncol(design), number = Inf)
  dm <- merge(myann,dm,by=0)
  dm <- dm[order(dm$P.Value),]
  rownames(dm) <- dm$Row.names
  dm$Row.names=NULL
  return(dm)
}

dm <- runlimma(mval=mval,design=design,myann=myann)

```

Reactome pathways were downloaded on the 14th Sept 2023 from MsigDB.

```{r,gsdl}

gs_entrez <- gmt_import("c2.cp.reactome.v2023.1.Hs.entrez.gmt")

gs_symbols <- gmt_import("c2.cp.reactome.v2023.1.Hs.symbols.gmt")

```


## Load RNA expression

From GEO https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE158420
GSE158420_counts.txt.gz

## DESEQ2


```{r,deseq1}

if (!file.exists("GSE158420_counts.txt.gz")) {
  download.file("https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE158420&format=file&file=GSE158420%5Fcounts%2Etxt%2Egz",
    destfile="GSE158420_counts.txt.gz")
}

x <- read.table("GSE158420_counts.txt.gz",row.names=NULL)
dim(x)
x <- aggregate(. ~ row.names,x,sum)
dim(x)

#RUFKM
head(x)

```

There are some gene name errors, so I will use HGNChelper to repair them.

```{r,gene_name_fix}

human=x$row.names
new.hgnc.table <- readRDS("new.hgnc.table.rds")
fix <- checkGeneSymbols(human,map=new.hgnc.table)
x$gene <- fix$Suggested.Symbol
x$row.names=NULL
x <- aggregate(. ~ gene,x,sum)
dim(x)
row.names(x)=x$gene
x$gene=NULL
head(x)

```

Now do DESeq2

```{r,deseq2}

patient <- head(as.character(unlist(lapply(1:ncol(x),function(i) {c(i,i)}))),ncol(x))
tumor <- rep(c(0,1),ncol(x)/2)
ss <- data.frame(patient,tumor)

dds <- DESeqDataSetFromMatrix(countData = x , colData=ss, design = ~ patient + tumor)
res <- DESeq(dds)
z <- results(res)
vsd <- vst(dds, blind=FALSE)
zz <-cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])

sig <- subset(dge,padj<0.05)
SIG = nrow(sig)
DN = nrow(subset(sig,log2FoldChange<0))
UP = nrow(subset(sig,log2FoldChange>0))
HEADER = paste("mRNA", SIG , "DEGs,", UP ,"upregulated,", DN, "downregulated")
# smear
plot(log2(dge$baseMean),dge$log2FoldChange,cex=0.6,cex.axis=1.2,cex.lab=1.3,
  xlab="log2 base mean",
  ylab="log2 fold change" ,pch=19,col="#838383")
points(log2(sig$baseMean),sig$log2FoldChange,cex=0.6,pch=19,col="red")
mtext((HEADER),cex=1.2)
top <- head(sig,20)
# volcano
plot(dge$log2FoldChange, -log2(dge$pvalue) ,cex=0.6, cex.lab=1.3,cex.axis=1.2,
 ,xlab="log2 fold change", ylab="-log2 p-value" ,pch=19,col="#838383")
points(sig$log2FoldChange, -log2(sig$pvalue),cex=0.6,pch=19,col="red")
mtext((HEADER),cex=1.2)
# top N gene heatmap
colCols <- rep(c(0,1),ncol(x)/2)
colCols <- gsub("0","gray",colCols)
colCols <- gsub("1","black",colCols)

colfunc <- colorRampPalette(c("blue", "white", "red"))
heatmap.2(  as.matrix(dge[1:50,c(7:ncol(dge))]), col=colfunc(25),scale="row",
  ColSideColors=colCols ,
  trace="none",margins = c(6,10), cexRow=.6, cexCol=.5, main="Top 50 genes")

```

RNA expression pathway enrichment with mitch.

```{r,mitch}

rna <- mitch_import(dge, DEtype="deseq2")

res <- mitch_calc(rna, genesets=gs_symbols, priority="effect",cores=16)

mres <- res$enrichment_result
msig <- subset(mres,p.adjustANOVA<0.05)

head(subset(mres,p.adjustANOVA<0.05 & s.dist>0),20) %>%
  kbl(caption="mRNA upregulated pathways") %>%
  kable_styling("hover",full_width=FALSE)

head(subset(mres,p.adjustANOVA<0.05 & s.dist<0),20) %>%
  kbl(caption="mRNA downregulated pathways") %>%
  kable_styling("hover",full_width=FALSE)

```

## Load methylation enrichment data and calculate overlap

Use Euler diagrams to find the similarity.
As medians are normally distributed about the zero, one would expect equal numbers of up and down-
methylated gene sets.


## Curate the annotation

Use all probes on the chip.

```{r,anno1}

anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
myann <- data.frame(anno[,c("UCSC_RefGene_Name","UCSC_RefGene_Group","Islands_Name","Relation_to_Island")])
gp <- myann[,"UCSC_RefGene_Name",drop=FALSE]
gp2 <- strsplit(gp$UCSC_RefGene_Name,";")
names(gp2) <- rownames(gp)
gp2 <- lapply(gp2,unique)
gt <- stack(gp2)
colnames(gt) <- c("gene","probe")
gt$probe <- as.character(gt$probe)
dim(gt)
str(gt)

#new.hgnc.table <- getCurrentHumanMap()
new.hgnc.table <- readRDS("new.hgnc.table.rds")
fix <- checkGeneSymbols(gt$gene,map=new.hgnc.table)
fix2 <- fix[which(fix$x != fix$Suggested.Symbol),]
length(unique(fix2$x))
gt$gene <- fix$Suggested.Symbol

```

Here we will look at parallel analysis of gene body and promoter.

The approach here is to generate two mapping tables of probes to genes based on the annotation which is split
into gene body and promoter (TSS).

gene body=gb

gene promoter (TSS) = gp

```{r,promoterbody}

# partition body and promoter data(probes)
dmg <- strsplit(dm$UCSC_RefGene_Name,";")

body <- lapply(1:nrow(dm),function(i) {
  a <- dm[i,"UCSC_RefGene_Group"]
  a <- unlist(strsplit(a,";"))
  dmg[[i]][a=="Body"]
})
names(body) <- rownames(dm)
body <- lapply(body,unique)
gb <- stack(body)
colnames(gb) <- c("gene","probe")
gb$probe <- as.character(gb$probe)
dim(gb)
str(gb)

fix <- checkGeneSymbols(gb$gene,map=new.hgnc.table)
fix2 <- fix[which(fix$x != fix$Suggested.Symbol),]
length(unique(fix2$x))
gb$gene <- fix$Suggested.Symbol

tss <- lapply(1:nrow(dm),function(i) {
  a <- dm[i,"UCSC_RefGene_Group"]
  a <- unlist(strsplit(a,";"))
  dmg[[i]][grep("TSS",a)]
})
names(tss) <- rownames(dm)
tss <- lapply(tss,unique)
gp <- stack(tss)
colnames(gp) <- c("gene","probe")
gp$probe <- as.character(gp$probe)
dim(gp)
str(gp)

fix <- checkGeneSymbols(gp$gene,map=new.hgnc.table)
fix2 <- fix[which(fix$x != fix$Suggested.Symbol),]
length(unique(fix2$x))
gp$gene <- fix$Suggested.Symbol

# import for promoters
dmp <- dm[which(rownames(dm) %in%  gp$probe),]
mp <- mitch_import(x=dmp,DEtype="limma",geneTable=gp)
dim(dmp)
dim(mp)

# import bodies
dmb <- dm[which(rownames(dm) %in%  gb$probe),]
mb <- mitch_import(x=dmb,DEtype="limma",geneTable=gb)
dim(dmb)
dim(mb)

# promoter
dmp_res <- mitch_calc(mp, genesets=gs_symbols, priority="effect",cores=16)
dmp_res_up <- subset(dmp_res$enrichment_result,s.dist>0 & p.adjustANOVA < 0.05)
dmp_res_dn <- subset(dmp_res$enrichment_result,s.dist<0 & p.adjustANOVA < 0.05)
nrow(dmp_res_up)
nrow(dmp_res_dn)
head(dmp_res_up,20)
head(dmp_res_dn,20)

# body
dmb_res <- mitch_calc(mb, genesets=gs_symbols, priority="effect",cores=16)
dmb_res_up <- subset(dmb_res$enrichment_result,s.dist>0 & p.adjustANOVA < 0.05)
dmb_res_dn <- subset(dmb_res$enrichment_result,s.dist<0 & p.adjustANOVA < 0.05)
nrow(dmb_res_up)
nrow(dmb_res_dn)
head(dmb_res_up,20)
head(dmb_res_dn,20)

# combined methylation
mm <- merge(mp,mb,by=0)
rownames(mm) <- mm$Row.names
mm$Row.names=NULL
colnames(mm) <- c("prom","body")
mm_res <- mitch_calc(mm, genesets=gs_symbols, priority="effect",cores=16)
mm_res_sig <- subset(mm_res$enrichment_result,p.adjustMANOVA < 0.01)
nrow(mm_res_sig)
head(mm_res_sig,20)
table(sign(mm_res_sig$s.prom) == sign(mm_res_sig$s.body) )
mm_res_sig[sign(mm_res_sig$s.prom) != sign(mm_res_sig$s.body),]

```

## RNA expression and methylation enrichment

```{r,rnaoverlap2}

# three way
mmrna <- merge(mm,rna,by=0)
rownames(mmrna) <- mmrna$Row.names
mmrna$Row.names = NULL
colnames(mmrna) <- c("prom","body","rna")
mmrnares <- mitch_calc(mmrna, genesets=gs_symbols, priority="effect",cores=16)
mmrnares_sig <- subset(mmrnares$enrichment_result, p.adjustMANOVA < 0.05)
head(mmrnares_sig,20)
table(sign(mmrnares_sig$s.prom) == sign(mmrnares_sig$s.body))
table(sign(mmrnares_sig$s.prom) == sign(mmrnares_sig$s.rna))
table(sign(mmrnares_sig$s.body) == sign(mmrnares_sig$s.rna))
mtop <- head(mmrnares_sig,30)
mtop <- mtop[,c("set","s.prom","s.body","s.rna")]
rownames(mtop) <- gsub("REACTOME_","",mtop$set)
rownames(mtop) <- gsub("_"," ",rownames(mtop))
mtop$set=NULL
colnames(mtop) <- c("promoter","body","RNA")
colfunc <- colorRampPalette(c("blue", "white", "red"))
heatmap.2( as.matrix(mtop) , trace="none",margins = c(6,25), col=colfunc(25), cexRow=.6, cexCol=1)

pdf("fig5c.pdf")
heatmap.2( as.matrix(mtop) , trace="none",margins = c(6,25), col=colfunc(25), cexRow=.65, cexCol=1)
dev.off()

# rna vs promoter
mprna <- merge(mp,rna,by=0)
rownames(mprna) <- mprna$Row.names
mprna$Row.names = NULL
colnames(mprna) <- c("prom","rna")
mprnares <- mitch_calc(mprna, genesets=gs_symbols, priority="effect",cores=16)
mprnares_sig <- subset(mprnares$enrichment_result, p.adjustMANOVA < 0.05)
head(mprnares_sig,20)
table(sign(mprnares_sig$s.prom) == sign(mmrnares_sig$s.rna))
mprnares_sig <- subset(mprnares$enrichment_result, p.adjustMANOVA < 0.05)
mylm <- lm(mprnares$enrichment_result$s.rna ~ mprnares$enrichment_result$s.prom)
summary(mylm)
mylm2 <- lm(mprnares_sig$s.rna ~ mprnares_sig$s.pr)
summary(mylm2)
plot(mprnares$enrichment_result$s.prom,mprnares$enrichment_result$s.rna,xlab="promoter",
  ylab="RNA", main="S dist",col="gray",pch=19)
abline(v=0,h=0,lty=2,lwd=2)
points(mprnares_sig$s.pr,mprnares_sig$s.rna,col="black",pch=19,cex=1.01)
abline(mylm,col="blue",lty=3,lwd=3)
abline(mylm2,col="red",lty=3,lwd=3)

# rna vs body
mbrna <- merge(mb,rna,by=0)
rownames(mbrna) <- mbrna$Row.names
mbrna$Row.names = NULL
colnames(mbrna) <- c("body","rna")
mbrnares <- mitch_calc(mbrna, genesets=gs_symbols, priority="effect",cores=16)
mbrnares_sig <- subset(mbrnares$enrichment_result, p.adjustMANOVA < 0.05)
head(mbrnares_sig,20)
table(sign(mbrnares_sig$s.body) == sign(mbrnares_sig$s.rna))
mbrnares_sig <- subset(mbrnares$enrichment_result, p.adjustMANOVA < 0.05)
mylm <- lm(mbrnares$enrichment_result$s.rna ~ mbrnares$enrichment_result$s.body)
summary(mylm)
mylm2 <- lm(mbrnares_sig$s.rna ~ mbrnares_sig$s.body)
summary(mylm2)
plot(mbrnares$enrichment_result$s.body,mbrnares$enrichment_result$s.rna,xlab="ex-promoter",
  ylab="RNA",main="S dist",col="gray",pch=19)
abline(v=0,h=0,lty=2,lwd=2)
points(mbrnares_sig$s.body,mbrnares_sig$s.rna,col="black",pch=19,cex=1.01)
abline(mylm,col="blue",lty=3,lwd=3)
abline(mylm2,col="red",lty=3,lwd=3)

```

Now look at GSAmeth.

```{r,rnagsa}

# body
sigup <- rownames(subset(dm,logFC > 0 & adj.P.Val < 0.05))
sigdn <- rownames(subset(dm,logFC < 0 & adj.P.Val < 0.05))
gsaup <- gsameth(sig.cpg=sigup, all.cpg=rownames(dm), collection=gs_entrez,
  array.type="EPIC", genomic.features = "Body")
rownames(gsaup) <- paste(rownames(gsaup),"UP")
gsadn <- gsameth(sig.cpg=sigdn, all.cpg=rownames(dm), collection=gs_entrez,
  array.type="EPIC", genomic.features = "Body")
rownames(gsadn) <- paste(rownames(gsadn),"DN")
gsig_up <- subset(gsaup,FDR<0.05)
gsig_dn <- subset(gsadn,FDR<0.05)
gsig <- rbind(gsig_up,gsig_dn)
nrow(gsig)
gsig_body <- gsig
gsig_body_up <- gsig_up
gsig_body_dn <- gsig_dn

# promoter
TSS = c("TSS200", "TSS1500")
gsaup <- gsameth(sig.cpg=sigup, all.cpg=rownames(dm), collection=gs_entrez,
  array.type="EPIC", genomic.features = TSS)
rownames(gsaup) <- paste(rownames(gsaup),"UP")
gsadn <- gsameth(sig.cpg=sigdn, all.cpg=rownames(dm), collection=gs_entrez,
  array.type="EPIC", genomic.features = TSS)
rownames(gsadn) <- paste(rownames(gsadn),"DN")
gsig_up <- subset(gsaup,FDR<0.05)
gsig_dn <- subset(gsadn,FDR<0.05)
gsig <- rbind(gsig_up,gsig_dn)
nrow(gsig)
gsig_promoter <- gsig
gsig_promoter_up <- gsig_up
gsig_promoter_dn <- gsig_dn

```

Now make a Euler plot for GSA meth-RNA comparison.

```{r,compare_rna_gsa}

rna_up <- subset(msig,s.dist>0)$set
rna_dn <- subset(msig,s.dist<0)$set

gsig_body_up <- rownames(gsig_body_up)
gsig_body_dn <- rownames(gsig_body_dn)

gsig_promoter_up <- rownames(gsig_promoter_up)
gsig_promoter_dn <- rownames(gsig_promoter_dn)

vg <- list("RNA up"=rna_up,"RNA dn"=rna_dn,
  "GSA body up"=gsig_body_up, "GSA body dn"=gsig_body_dn,
  "GSA promoter up"=gsig_promoter_up, "GSA prom dn"=gsig_promoter_dn)

plot(euler(vg),quantities=TRUE)

pdf("fig5a.pdf")
plot(euler(vg),quantities=TRUE)
dev.off()

```

Now make a Euler plot for LAM-RNA comparison.

```{r,compare_rna_mitch}

dmp_res_up <- dmp_res_up$set
dmp_res_dn <- dmp_res_dn$set

dmb_res_up <- dmb_res_up$set
dmb_res_dn <- dmb_res_dn$set

vm <- list("RNA up"=rna_up,"RNA dn"=rna_dn,
  "LAM body up"=dmb_res_up, "LAM body dn"=dmb_res_dn,
  "LAM promoter up"=dmp_res_up, "LAM prom dn"=dmp_res_dn)

plot(euler(vm),quantities=TRUE)

pdf("fig5b.pdf")
plot(euler(vm),quantities=TRUE)
dev.off()

```

## Conclusion

Although there were some differentially methylated pathways analysed with different methods,
they were not concordant with differential expression.

## Session information

```{r,save}

save.image("GSE158422_rna_int.Rdata")

sessionInfo()

```
