---
title: "GSE158422: DNA methylation in Lung squamous cell carcinoma - gene methylation enrichment analysis - sensitivity analysis"
author: "The GMEA team"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    fig_width: 7
    fig_height: 7
theme: cosmo
---

## Introduction

In the previous analysis, it was shown that the "AL" (aggregation-limma-enrichment) method
yielded good performance with a cohort of cancer-normal tissue pairs.

In this piece of work, we want to ascertain the sensitivity of this approach as compared to
existing methods such as gsameth and ebGSEA.

To achieve this we will analyse with the full 37 patient sample pairs, and repeatedly downsample
the dataset down to n=3 patients.
What we expect is that the "AL" method could have a higher degree of recall as compared to these
other approaches.
We expect that "AL" could identify gene sets with larger effect sizes (enrichment scores) using
fewer samples.
This means that for a given set of datasets, this approach could yield more findings, or
alternatively if pathway analysis is the primary outcome of a study, it need not be as large which
can save money and difficulty with establishing large sample collections.

## Requirements

This analysis was run on a 16C/32T computer with 128GB RAM running at 4.2 GHz.
RAM usage can be moderated by reducing the parallel core count.

```{r,packages}

suppressPackageStartupMessages({
  library("limma")
  library("eulerr")
  library("IlluminaHumanMethylation450kmanifest")
  library("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
  library("tictoc")
  library("mitch")
  library("kableExtra")
  library("beeswarm")
  library("missMethyl")
})

CORES=24

```

## Load data

* annotations

* probe sets

* gene sets

* design matrix

* mval matrix

```{r,annotation}

anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
myann <- data.frame(anno[,c("UCSC_RefGene_Name","Regulatory_Feature_Group","Islands_Name","Relation_to_Island")])

gp <- myann[,"UCSC_RefGene_Name",drop=FALSE]
gp2 <- strsplit(gp$UCSC_RefGene_Name,";")
names(gp2) <- rownames(gp)
sets <- split(rep(names(gp2), lengths(gp2)), unlist(gp2))

summary(unlist(lapply(sets,length)))

genesets <- gmt_import("https://ziemann-lab.net/public/gmea_prototype/ReactomePathways.gmt")

if (!file.exists("GSE158422_design.rds")) {
  download.file("https://ziemann-lab.net/public/gmea_prototype/GSE158422_design.rds", "GSE158422_design.rds")
}
design <- readRDS("GSE158422_design.rds")

if (!file.exists("GSE158422_design.rds")) {
 download.file("https://ziemann-lab.net/public/gmea_prototype/GSE158422_mx.rds","GSE158422_mx.rds")
}
mval <- readRDS("GSE158422_mx.rds")

mval_means <- list("tumor"=matrix(colMeans(mval),ncol=2)[,1],"normal"=matrix(colMeans(mval),ncol=2)[,2])
boxplot(mval_means, main="mean probe methylation mval",cex=0,col="white")
beeswarm(mval_means,add=TRUE,pch=1,cex=1.5)
lapply(mval_means,mean)

```

We will assume mean probe methylation is a proxy measure for global methylation.
The boxplot appears to show higher values in tumour samples, based on median and quartiles.
However the mean values are actually lower in tumour samples compared to normal.

Reactome pathways were downloaded on the 14th Sept 2023 from MsigDB.

```{r,gsdl}

gs_entrez <- gmt_import("c2.cp.reactome.v2023.1.Hs.entrez.gmt")

gs_symbols <- gmt_import("c2.cp.reactome.v2023.1.Hs.symbols.gmt")

```

# AL GMEA: all probes

Start with curating the sample info and design matrix.

```{r,samplesheet}

sex <- as.data.frame(design)$sex
tumor <- as.data.frame(design)$tumor
patient <- as.character(unlist(lapply(1:ncol(mval),function(i) {c(i,i)})))
patient <- head(patient,ncol(mval))
design <- model.matrix(~ patient + tumor )

```

## Functions

```{r,functions}

# magg is a general function that aggregates probe methylation measurements by gene
# in this variation it aggregates by MEAN, which performs better than median
magg <- function(mval,myann,cores=1){
  gn <- unique(unlist(strsplit( myann$UCSC_RefGene_Name ,";")))
  gnl <- strsplit( myann$UCSC_RefGene_Name ,";")
  gnl <- mclapply(gnl,unique,mc.cores=cores)
  myann$gnl <- gnl
  keep <- rownames(subset(myann,UCSC_RefGene_Name!=""))
  mx <- mval[rownames(mval) %in% keep,]
  mymed <- function(g) {
    probes <- rownames(myann[grep(g,myann$gnl),])
    rows <- which(rownames(mx) %in% probes)
    if ( length(rows) > 1 ) {
      b <- mx[rows,]
      med <- apply(b,2,mean)
      med <- matrix(med,nrow=1)
      colnames(med) <- colnames(b)
      rownames(med) <- g
      return(med)
    }
  }
  med <- mclapply(gn,mymed,mc.cores=cores)
  med <- med[lapply(med,length)>0]
  medf <- do.call(rbind,med)
  return(medf)
}

# chragg is a specialised function that performs the magg routine on chromosomes in parallel
chragg <- function(mval,myann,cores=1){
  annodf <- as.data.frame(anno)
  keep <- rownames(subset(myann,UCSC_RefGene_Name!=""))
  mx <- mval[rownames(mval) %in% keep,]
  chrs <- unique(anno$chr)
  myorder <- unlist(lapply(chrs,function(mychr) { nrow( annodf[annodf$chr==mychr,] ) } ))
  chrs <- chrs[order(-myorder)]
  leadercores <- floor(sqrt(cores))
  workercores <- ceiling(sqrt(cores))
  chrmedf <- mclapply(chrs,function(chr) {
    chrfrag <- annodf[annodf$chr==chr,]
    chrprobes <-rownames(chrfrag)
    chrmx <- mx[rownames(mx) %in% chrprobes,]
    chranno <- myann[rownames(myann) %in% chrprobes,]
    chrmedf <- magg(mval=chrmx,myann=chranno,cores=workercores)
    return(chrmedf)
  },mc.cores=leadercores)
  medf <- do.call(rbind, chrmedf)
  return(medf)
}

# agglimma is a helper function that runs limma analysis on the aggregated methylation data
# and fixes any duplicated gene names - most significant one is retained
agglimma <- function(medf,design) {
  fit.reduced <- lmFit(medf,design)
  fit.reduced <- eBayes(fit.reduced)
  dmagg <- topTable(fit.reduced,coef=ncol(design), number = Inf)
  nondup <- !duplicated(dmagg$ID)
  dmagg <- dmagg[nondup,]
  rownames(dmagg) <- dmagg$ID
  dmagg$ID = NULL
  return(dmagg)
}

# ttenrich is a function designed to conduct analysis of differential methylation.
# It takes the t-statistics from limma for a geneset and determines whether this gene
# set has a distribution different to the null (zero) using a one-sample two-sided t-test.
ttenrich <- function(m,genesets,cores=1) {
  res <- mclapply( 1:length(genesets), function(i) {
    gs <- genesets[i]
    name <- names(gs)
    n_members <- length(which(rownames(m) %in% gs[[1]]))
    if ( n_members > 4 ) {
      tstats <- m[which(rownames(m) %in% gs[[1]]),]
      myn <- length(tstats)
      mymean <- mean(tstats)
      mymedian <- median(tstats)
      wt <- t.test(tstats)
      res <- c(name,myn,mymean,mymedian,wt$p.value)
    }
  } , mc.cores = cores)
  res_df <- do.call(rbind, res)
  rownames(res_df) <- res_df[,1]
  res_df <- res_df[,-1]
  colnames(res_df) <- c("n_genes","t_mean","t_median","pval")
  tmp <- apply(res_df,2,as.numeric)
  rownames(tmp) <- rownames(res_df)
  res_df <- tmp
  res_df <- as.data.frame(res_df)
  res_df <- res_df[order(res_df$pval),]
  res_df$logp <- -log10(res_df$pval )
  res_df$fdr <- p.adjust(res_df$pval,method="fdr")
  res_df[order(abs(res_df$pval)),]
  return(res_df)
}

# runmitch is a helper function that runs mitch on a 1 column dataframe and just returns
# the differential table
runmitch <- function(m,genesets,cores=1) {
  mres <- mitch_calc(m,genesets,minsetsize=5,cores=cores)
  mres <- mres$enrichment_result
  rownames(mres) <- mres$set
  mres$set=NULL
  return(mres)
}

```

Run AL method on methylation data.
Record timings.

```{r,rungmea}

# aggregation step
tic()
medf <- chragg(mval,myann,cores=CORES)
toc() #51.5 sec

medf_means <- list("tumor"=matrix(colMeans(medf),ncol=2)[,1],"normal"=matrix(colMeans(medf),ncol=2)[,2])
boxplot(medf_means, main="mean gene methylation mval",cex=0,col="white")
beeswarm(medf_means,add=TRUE,pch=1,cex=1.5)

lapply(medf_means,mean)

# differential analysis
tic()
dmagg <- agglimma(medf,design)
toc() #0.46 sec

hist(dmagg$t)

head(dmagg, 30) %>%
  kbl(caption="limma gene methylation results") %>%
  kable_paper("hover", full_width = F)

# get t-stats
m <- as.data.frame(dmagg$t)
rownames(m) <- rownames(dmagg)
colnames(m) <- "t"

# enrichment test
tic()
alres <- ttenrich(m=m,genesets=gs_symbols,cores=CORES)
toc() #1.1 sec

hist(alres$t_mean)

alsig <- subset(alres,fdr<0.05)

nrow(alsig)

nrow(subset(alsig,t_median>0))

head(alsig[order(-alsig$t_median),],20) %>%
  kbl(caption="hypermethylated pathways") %>%
  kable_paper("hover", full_width = F)

nrow(subset(alsig,t_median<0))

head(alsig[order(alsig$t_median),],20) %>%
  kbl(caption="hypomethylated pathways") %>%
  kable_paper("hover", full_width = F)

```

## Downsample AL test

```{r,al_down1,fig.height=10}

#medf already exists

SAMPLESIZES=c(2,3,6,10,16,22,30)

ssres <- lapply(SAMPLESIZES,function(n) {
  SEEDS <- seq(100,5000,100)
  dsres <- mclapply(SEEDS, function(SEED) {
    set.seed(SEED)
    mysample <- sample(1:37,n,replace=FALSE)*2
    mysample <- c(mysample,(mysample-1))
    mysample <- mysample[order(mysample)]
    design2 <- design[mysample,]
    design2 <- design2[,colSums(design2)>0]
    medf2 <- medf[,mysample]
    dmagg2 <- agglimma(medf2,design2)
    m2 <- as.data.frame(dmagg2$t)
    rownames(m2) <- rownames(dmagg2)
    colnames(m2) <- "t"
    alres2 <- ttenrich(m=m2,genesets=gs_symbols,cores=CORES)
    head(alres2)
    alsig2 <- subset(alres2,fdr<0.05)
    TOT=nrow(alsig)
    TP=length(which(rownames(alsig2) %in% rownames(alsig)))
    FP=length(which(!rownames(alsig2) %in% rownames(alsig)))
    TPES=mean(abs(alsig[which(rownames(alsig) %in% rownames(alsig2)),"t_mean"]))
    RES=c("TOT"=TOT,"TP"=TP,"FP"=FP,"TPES"=TPES)
    RES
  },mc.cores=2)
  do.call(rbind,dsres)
})

par(mfrow=c(3,1))

dots <- lapply(ssres,function(x) { x[,"TP"] })
boxplot(dots,names=SAMPLESIZES,col="white",ylab="no. true positives",xlab="sample size",cex=0,main="GMEA") ; grid()
beeswarm(dots,add=TRUE,pch=1,cex=0.6)
abline(h= ssres[[1]][1,1])

dots <- lapply(ssres,function(x) { x[,"FP"] })
boxplot(dots,names=SAMPLESIZES,col="white",ylab="no. false positives",xlab="sample size",cex=0) ; grid()
beeswarm(dots,add=TRUE,pch=1,cex=0.6)

dots <- lapply(ssres,function(x) { x[,"TP"] })
TPdots <- do.call(rbind,dots)
dots <- lapply(ssres,function(x) { x[,"FP"] })
FPdots <- do.call(rbind,dots)
ratio <- t(FPdots / (TPdots + FPdots))
ratio <- lapply(1:ncol(ratio),function(i) {ratio[,i]})
boxplot(ratio,names=SAMPLESIZES,col="white",ylab="proportion FP",xlab="sample size",cex=0) ; grid()
beeswarm(ratio,add=TRUE,pch=1,cex=0.6)

```

## ALM aggregate limma mitch

Try a competitive test.

```{r,compet}

mres <- mitch_calc(x=m,minsetsize=5,genesets=gs_symbols,cores=32,priority="effect")
almsig <- subset(mres$enrichment_result,p.adjustANOVA<0.05)
rownames(almsig) <- almsig$set
alm$set=NULL

# hypermethylated
mres_up <- subset(mres$enrichment_result,p.adjustANOVA<0.05 & s.dist > 0)
nrow(mres_up)
head(mres_up,20) %>%
  kbl(caption="hypermethylated pathways with competitive test") %>%
  kable_paper("hover", full_width = F)

# hypomethylated
mres_dn <- subset(mres$enrichment_result,p.adjustANOVA<0.05 & s.dist < 0)
nrow(mres_dn)
head(mres_dn,20) %>%
  kbl(caption="hypomethylated pathways with competitive test") %>%
  kable_paper("hover", full_width = F)

```

These results look more in line with cancer pathogenesis.

Downsample alm method

```{r,almdown}

SAMPLESIZES=c(2,3,6,10,16,22,30)

mmres <- lapply(SAMPLESIZES,function(n) {
  SEEDS <- seq(100,5000,100)
  dsres <- mclapply(SEEDS, function(SEED) {
    set.seed(SEED)
    mysample <- sample(1:37,n,replace=FALSE)*2
    mysample <- c(mysample,(mysample-1))
    mysample <- mysample[order(mysample)]
    design2 <- design[mysample,]
    design2 <- design2[,colSums(design2)>0]
    medf2 <- medf[,mysample]
    dmagg2 <- agglimma(medf2,design2)
    m2 <- as.data.frame(dmagg2$t)
    rownames(m2) <- rownames(dmagg2)
    colnames(m2) <- "t"
    almres2 <- runmitch(m=m2,genesets=gs_symbols,cores=CORES)
    head(almres2)
    almsig2 <- subset(almres2,p.adjustANOVA<0.05)
    TOT=nrow(almsig)
    TP=length(which(rownames(almsig2) %in% rownames(almsig)))
    FP=length(which(!rownames(almsig2) %in% rownames(almsig)))
    TPES=1
    RES=c("TOT"=TOT,"TP"=TP,"FP"=FP,"TPES"=TPES)
    RES
  },mc.cores=2)
  do.call(rbind,dsres)
})

par(mfrow=c(3,1))

dots <- lapply(mmres,function(x) { x[,"TP"] })
boxplot(dots,names=SAMPLESIZES,col="white",ylab="no. true positives",xlab="sample size",cex=0,main="ALM") ; grid()
beeswarm(dots,add=TRUE,pch=1,cex=0.6)
abline(h=mmres[[1]][1,1])

dots <- lapply(mmres,function(x) { x[,"FP"] })
boxplot(dots,names=SAMPLESIZES,col="white",ylab="no. false positives",xlab="sample size",cex=0) ; grid()
beeswarm(dots,add=TRUE,pch=1,cex=0.6)

dots <- lapply(mmres,function(x) { x[,"TP"] })
TPdots <- do.call(rbind,dots)
dots <- lapply(mmres,function(x) { x[,"FP"] })
FPdots <- do.call(rbind,dots)
ratio <- t(FPdots / (TPdots + FPdots))
ratio <- lapply(1:ncol(ratio),function(i) {ratio[,i]})
boxplot(ratio,names=SAMPLESIZES,col="white",ylab="proportion FP",xlab="sample size",cex=0) ; grid()
beeswarm(ratio,add=TRUE,pch=1,cex=0.6)

```

# GSAMETH

Run GSA meth on the full dataset.

```{r,rungsameth1}

# limma
runlimma <- function(mval,design,myann) {
  fit.reduced <- lmFit(mval,design)
  fit.reduced <- eBayes(fit.reduced)
  summary(decideTests(fit.reduced))
  dm <- topTable(fit.reduced,coef=4, number = Inf)
  dm <- merge(myann,dm,by=0)
  dm <- dm[order(dm$P.Value),]
  rownames(dm) <- dm$Row.names
  dm$Row.names=NULL
  return(dm)
}

dm <- runlimma(mval,design,myann)

hist(dm$t,breaks=100,xlim=c(-10,10)) ; grid()

head(dm, 10) %>% kbl() %>% kable_paper("hover", full_width = F)

sigup <- rownames(subset(dm,logFC > 0 & adj.P.Val < 0.05))
sigdn <- rownames(subset(dm,logFC < 0 & adj.P.Val < 0.05))

length(sigup)
length(sigdn)

tic()
gsaup <- gsameth(sig.cpg=sigup, all.cpg=rownames(dm), collection=gs_entrez)
gsadn <- gsameth(sig.cpg=sigdn, all.cpg=rownames(dm), collection=gs_entrez)
toc() # 12.9 sec

gsaup <- gsaup[order(gsaup$P.DE),]
gsadn <- gsadn[order(gsadn$P.DE),]

gsig_up <-  subset(gsaup,FDR<0.05)
gsig_up %>% kbl(caption="Hypermethylated pathways") %>% kable_paper("hover", full_width = F)
gsig_dn <-  subset(gsadn,FDR<0.05)
gsig_dn %>% kbl(caption="Hypomethylated pathways") %>% kable_paper("hover", full_width = F)
gsig <- rbind(gsig_up,gsig_dn)

```

Only 4 hyper and 1 hypomethylated gene set were identified.
This is surprising as there were 1299 hypermethylated probes corresponding to 808 genes, and
1016 hypomethylated probes corresponding to 657 genes.

## GSAmeth Downsample

```{r,gsamethdownsample,fig.height=10}

SAMPLESIZES=c(2,3,6,10,16,22,30)

gsres <- lapply(SAMPLESIZES,function(n) {
  SEEDS <- seq(100,5000,100)
  dsres <- mclapply(SEEDS, function(SEED) {
    set.seed(SEED)
    mysample <- sample(1:37,n,replace=FALSE)*2
    mysample <- c(mysample,(mysample-1))
    mysample <- mysample[order(mysample)]
    design2 <- design[mysample,]
    design2 <- design2[,colSums(design2)>0]
    #mval downsample
    mval2 <- mval[,mysample]
    dm2 <- runlimma(mval2,design2,myann)
    sigup2 <- rownames(subset(dm2,adj.P.Val<0.05 & logFC>0))
    sigdn2 <- rownames(subset(dm2,adj.P.Val<0.05 & logFC<0))
    #gsamet
    gsaup2 <- gsameth(sig.cpg=sigup2, all.cpg=rownames(dm2), collection=gs_entrez)
    gsadn2 <- gsameth(sig.cpg=sigdn2, all.cpg=rownames(dm2), collection=gs_entrez)
    gsig_up2 <- subset(gsaup2,FDR<0.05)
    gsig_dn2 <- subset(gsadn2,FDR<0.05)
    gsig2 <- rbind(gsig_up2,gsig_dn2)
    TOT=nrow(gsig2)
    TP=length(which(rownames(gsig2) %in% rownames(gsig)))
    FP=length(which(!rownames(gsig2) %in% rownames(gsig)))
    TPES=0
    RES=c("TOT"=TOT,"TP"=TP,"FP"=FP,"TPES"=TPES)
    RES
  },mc.cores=8)
  do.call(rbind,dsres)
})

mx <- matrix(rep(0,200),ncol=4)
colnames(mx) <- c("TOT","TP","FP","TPES")

# fix empty matrix
gsres <- lapply(gsres,function(x) {
  if ( typeof(x) == "character" ) {
    mx
  } else {
    x
  }
} )

par(mfrow=c(3,1))

dots <- lapply(gsres,function(x) { x[,"TP"] })
boxplot(dots,names=SAMPLESIZES,col="white",ylab="no. true positives",xlab="sample size",cex=0,main="GSAmeth")
beeswarm(dots,add=TRUE,pch=1,cex=0.6)
abline(h=gsres[[1]][1,1])

dots <- lapply(gsres,function(x) { x[,"FP"] })
boxplot(dots,names=SAMPLESIZES,col="white",ylab="no. false positives",xlab="sample size",cex=0)
beeswarm(dots,add=TRUE,pch=1,cex=0.6)

dots <- lapply(gsres,function(x) { x[,"TP"] })
TPdots <- do.call(rbind,dots)
dots <- lapply(gsres,function(x) { x[,"FP"] })
FPdots <- do.call(rbind,dots)
ratio <- t(FPdots / (TPdots + FPdots))
ratio <- lapply(1:ncol(ratio),function(i) {ratio[,i]})
boxplot(ratio,names=SAMPLESIZES,col="white",ylab="proportion FP",xlab="sample size",cex=0) ; grid()
beeswarm(ratio,add=TRUE,pch=1,cex=0.6)

par(mfrow=c(1,1))

```

## Overlap

Use Euler diagrams to find the similarity.
As medians are normally distributed about the zero, one would expect equal numbers of up and down-
methylated gene sets.
Therefore the t-test based approach could be considered better.

```{r,ol1}

alsig_up <- rownames(subset(alsig,t_median>0))
alsig_dn <- rownames(subset(alsig,t_median<0))

msig_up <- subset(mres$enrichment_result,p.adjustANOVA<0.05 & s.dist > 0)$set
msig_dn <- subset(mres$enrichment_result,p.adjustANOVA<0.05 & s.dist < 0)$set

gsig_up <-  rownames(subset(gsaup,FDR<0.05))
gsig_dn <-  rownames(subset(gsadn,FDR<0.05))

v1 <- list("AL up"=alsig_up,"AL dn"=alsig_dn,
  "mitch up"=msig_up,"mitch dn"=msig_dn,
  "GSAmeth up"=gsig_up,"GSAmeth dn"=gsig_dn)

plot(euler(v1),quantities = TRUE)

```

## Session information

```{r,save}

save.image("GSE158422_sensitivity.Rdata")

sessionInfo()

```
