---
title: "GOmeth test"
author: "hcy_meth meta-analysis group"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_float: true
    code_folding: hide
    fig_width: 7
    fig_height: 7
theme: cosmo
---

## Introduction

Here we will run test enrichment analysis approaches for understanding the effect of elevated
homocysteine on blood methylomes.
Homocysteine is thought to be an inhibitor of methyltransferases and hyperhomocysteine is
associated with overall reduced genomic methylation.
Previously we analysed the B-PROOF study data which enabled us to identify probes whose methylation correlatesor anti-correlated with homocysteine level.

The two enrichment analysis approaches we will use are:

* `gsameth`: an ORA approach that addressed the biases inherent with infinium array analysis.

* `gmea`: an experimental application of functional class scoring for infinium analysis.

```{r,libs}

suppressPackageStartupMessages({
  library("missMethyl")
  library("kableExtra")
  library("mitch")
  library("limma")
  library("eulerr")
  library("IlluminaHumanMethylation450kmanifest")
  library("IlluminaHumanMethylation450kanno.ilmn12.hg19")
  library("tictoc")
})

CORES=12

```

Load the data.

```{r,load}

dm <- read.table("dma3a.tsv")
rownames(dm) <- dm$Row.names
dm$Row.names=NULL

head(dm, 10) %>% kbl() %>% kable_paper("hover", full_width = F)

```

Reactome pathways were downloaded on the 14th Sept 2023 from MsigDB.

```{r,gsdl}

gs_entrez <- gmt_import("c2.cp.reactome.v2023.1.Hs.entrez.gmt")

gs_symbols <- gmt_import("c2.cp.reactome.v2023.1.Hs.symbols.gmt")

```

## ORA with GSAMETH - parth of the MissMethyl package

Now perform ORA analysis using gsameth with the top 500 probes.


```{r,gsameth1}

sigup <- rownames(head(subset(dm,logFC>0),1000))
sigdn <- rownames(head(subset(dm,logFC<0),1000))

tic()
gsaup <- gsameth(sig.cpg=head(sigup,500), all.cpg=rownames(dm), collection=gs_entrez)
gsadn <- gsameth(sig.cpg=head(sigdn,500), all.cpg=rownames(dm), collection=gs_entrez)
toc()

gsaup <- gsaup[order(gsaup$P.DE),]
head(gsaup,10) %>% kbl(caption="Top hypermethylated pathways") %>% kable_paper("hover", full_width = F)

gsadn <- gsadn[order(gsadn$P.DE),]
head(gsadn,10) %>% kbl(caption="Top hypomethylated pathways") %>% kable_paper("hover", full_width = F)

gsig_up <-  subset(gsaup,FDR<0.1)
gsig_dn <-  subset(gsadn,FDR<0.1)

```

## GMEA

Now with GMEA.
There are two approaches:

1. limma-agg-1stt (la1).
In this approach, the limma test is conducted on probes, followed by aggregation of the differential
methylation t-scores. Each gene gets a median t-stat value which is evaluated downstream by a 1 sample
t-test.

2. agg-limma-1stt (al1).
Here, the probe beta values for each gene are aggregated to a median and then limma is run on these
values.
To determine gene set enrichment, a 1-sample t-test is conducted.


Start with gathering the annotation set.

```{r,anno}

ann450k <- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
myann <- data.frame(ann450k[,c("UCSC_RefGene_Name","Regulatory_Feature_Group")])
promoters <- grep("Prom",myann$Regulatory_Feature_Group)

```

Check histogram of limma t-statistics.

```{r,hist}

dm <- dm[,c("UCSC_RefGene_Name","t")]
hist(dm$t,breaks=seq(from=-6,to=6,by=1)) ; grid()

```

## LA: limma-aggregate approach

Aggregate probe t-scores to gene level scores and conduct a 1-sample t-test on the gene level.

The object generated has two parts:

1. `df`: a simple table of the gene name and the median tvalue

2. `gme_res_df`: results of 1-sample Wilcox test.

```{r,agg_tt}

agg2 <- function(dm,cores=1) {
  gn <- unique(unlist(strsplit( dm$UCSC_RefGene_Name ,";")))
  gnl <- strsplit( dm$UCSC_RefGene_Name ,";")
  gnl <- mclapply(gnl,unique,mc.cores=cores)
  dm$UCSC_RefGene_Name <- gnl
  l <- mclapply(1:nrow(dm), function(i) {
    a <- dm[i,]
    len <- length(a[[1]][[1]])
    tvals <- as.numeric(rep(a["t"],len))
    genes <- a[[1]][[1]]
    data.frame(genes,tvals)
  },mc.cores=cores)
  df <- do.call(rbind,l)
  keep <- names(which(table(df$genes)>1))
  df <- df[df$genes %in% keep,]
  gn <- unique(df$genes)
  gme_res <- lapply( 1:length(gn), function(i) {
    g <- gn[i]
    tstats <- df[which(df$genes==g),"tvals"]
    myn <- length(tstats)
    mymean <- mean(tstats)
    mymedian <- median(tstats)
    if ( length(tstats) ) > 2 {
      ttest <- t.test(tstats)
      pval <- ttest$p.value
    } else {
      pval = 1
    }
    res <- c("gene"=g,"nprobes"=myn,"mean"=mymean,"median"=mymedian,
      "pval"=ttest$p.value)
  } )
  gme_res_df <- do.call(rbind, gme_res)
  rownames(gme_res_df) <- gme_res_df[,1]
  gme_res_df <- gme_res_df[,-1]
  tmp <- apply(gme_res_df,2,as.numeric)
  rownames(tmp) <- rownames(gme_res_df)
  gme_res_df <- as.data.frame(tmp)
  gme_res_df$sig <- -log10(gme_res_df[,4])
  gme_res_df <- gme_res_df[order(-gme_res_df$sig),]
  gme_res_df$`fdr(sc)` <- p.adjust(gme_res_df$`p-value(sc)`)
  out <- list("df"=df,"gme_res_df"=gme_res_df)
  return(out)
}

tic()
dmagg2 <- agg2(dm,cores=CORES)
toc()

head(dmagg2$gme_res_df) %>%
  kbl(caption="top gmea genes (t-test)") %>%
  kable_paper("hover", full_width = F)

```

Check that the distribution is even.

```{r,tmeddist}

hist(dmagg2$gme_res_df$median,breaks=30) ; grid()

```

1-sample t-test for gene set enrichment.

```{r,t_enrichment_test1}

ttenrich <- function(m,genesets,cores=1) {
  res <- mclapply( 1:length(genesets), function(i) {
    gs <- genesets[i]
    name <- names(gs)
    n_members <- length(which(rownames(m) %in% gs[[1]]))
    if ( n_members > 4 ) {
      tstats <- m[which(rownames(m) %in% gs[[1]]),]
      myn <- length(tstats)
      mymean <- mean(tstats)
      mymedian <- median(tstats)
      wt <- t.test(tstats)
      res <- c(name,myn,mymean,mymedian,wt$p.value)
    }
  } , mc.cores = cores)
  res_df <- do.call(rbind, res)
  rownames(res_df) <- res_df[,1]
  res_df <- res_df[,-1]
  colnames(res_df) <- c("n_genes","t_mean","t_median","p.value")
  tmp <- apply(res_df,2,as.numeric)
  rownames(tmp) <- rownames(res_df)
  res_df <- tmp
  res_df <- as.data.frame(res_df)
  res_df <- res_df[order(res_df$p.value),]
  res_df$logp <- -log10(res_df$p.value )
  res_df$fdr <- p.adjust(res_df$p.value,method="fdr")
  res_df[order(abs(res_df$p.value)),]
  return(res_df)
}

# need to get the median column before analysis
m <- dmagg2$gme_res_df[,"median",drop=FALSE]

tic()
lares <- ttenrich(m=m,genesets=gs_symbols)
toc()

head(lares,30) %>%
  kbl(caption = "Top significant genesets using LA approach") %>%
  kable_paper("hover", full_width = F)

head(lares[order(-abs(lares$t_median)),],30) %>%
  kbl(caption = "Top effect size genesets using LA approach") %>%
  kable_paper("hover", full_width = F)

lasig <- subset(lares,fdr<0.05)
nrow(lasig)

head(lasig[order(-abs(lasig$t_median)),],30) %>%
  kbl(caption = "Top effect size genesets after FDR filtering using LA approach") %>%
  kable_paper("hover", full_width = F)

plot(lares$t_median,-log10(lares$p.value),pch=19,col="gray")
grid()
points(lasig$t_median,-log10(lasig$p.value),pch=19,col="red")

```

## AL: Aggregate limma t-test

Determine the median of probe values

```{r,med1}

gn <- unique(unlist(strsplit( myann$UCSC_RefGene_Name ,";")))
gnl <- strsplit( myann$UCSC_RefGene_Name ,";")
gnl <- mclapply(gnl,unique,mc.cores=CORES)
myann$gnl <- gnl

mymed <- function(g) {
  probes <- rownames(myann[grep(g,myann$gnl),])
  rows <- which(rownames(mx) %in% probes)
  if ( length(rows) > 1 ) {
    b <- mx[rows,]
    med <- apply(b,2,median)
    med <- matrix(med,nrow=1)
    colnames(med) <- colnames(b)
    rownames(med) <- g
    return(med)
  }
}

mx <- readRDS("dm3_mx.Rds")
design <- readRDS("dm3_design.Rds")

med <- mclapply(gn,mymed,mc.cores=CORES)
med <- med[lapply(med,length)>0]
medf <- do.call(rbind,med)

```

Then do a limma test on median gene methylation.
See if it is more robust than other approaches.

```{r,limma1}

fit.reduced <- lmFit(medf,design)
fit.reduced <- eBayes(fit.reduced)
summary(decideTests(fit.reduced))
magg <- topTable(fit.reduced,coef=4, number = Inf)

head(magg, 30) %>%
  kbl(caption="limma gene methylation results") %>%
  kable_paper("hover", full_width = F)

```

## 1-sample t-test

```{r,ttest}

ttenrich <- function(m,genesets,cores=1) {
  res <- mclapply( 1:length(genesets), function(i) {
    gs <- genesets[i]
    name <- names(gs)
    n_members <- length(which(rownames(m) %in% gs[[1]]))
    if ( n_members > 4 ) {
      tstats <- m[which(rownames(m) %in% gs[[1]]),]
      myn <- length(tstats)
      mymean <- mean(tstats)
      mymedian <- median(tstats)
      wt <- t.test(tstats)
      res <- c(name,myn,mymean,mymedian,wt$p.value)
    }
  } , mc.cores = cores)
  res_df <- do.call(rbind, res)
  rownames(res_df) <- res_df[,1]
  res_df <- res_df[,-1]
  colnames(res_df) <- c("n_genes","t_mean","t_median","p.value")
  tmp <- apply(res_df,2,as.numeric)
  rownames(tmp) <- rownames(res_df)
  res_df <- tmp
  res_df <- as.data.frame(res_df)
  res_df <- res_df[order(res_df$p.value),]
  res_df$logp <- -log10(res_df$p.value )
  res_df$fdr <- p.adjust(res_df$p.value,method="fdr")
  res_df[order(abs(res_df$p.value)),]
  return(res_df)
}

m <- as.data.frame(magg$t)
rownames(m) <- rownames(magg)
colnames(m) <- "t"

alres <- ttenrich(m=m,genesets=gs_symbols,cores=CORES)

alsig <- subset(alres,fdr<0.05)

nrow(alsig)

nrow(subset(alsig,t_median>0))

head(alsig[order(-alsig$t_median),],20) %>%
  kbl(caption="hypermethylated pathways") %>%
  kable_paper("hover", full_width = F)

nrow(subset(alsig,t_median<0))

head(alsig[order(alsig$t_median),],20) %>%
  kbl(caption="hypomethylated pathways") %>%
  kable_paper("hover", full_width = F)

```

## Overlap

Use Euler diagrams to find the similarity.
As medians are normally distributed about the zero, one would expect equal numbers of up and down-
methylated gene sets.
Therefore the t-test based approach could be considered better.

```{r,ol1}

lasig_up <- rownames(subset(lasig,t_median>0))
lasig_dn <- rownames(subset(lasig,t_median<0))

alsig_up <- rownames(subset(alsig,t_median>0))
alsig_dn <- rownames(subset(alsig,t_median<0))

v1 <- list("LA up"=lasig_up, "LA dn"=lasig_dn,
  "AL up"=alsig_up,"AL dn"=alsig_dn,
  "GSAmeth up"=gsig_up,"GSAmeth dn"=gsig_dn)

plot(euler(v1),quantities = TRUE)

```

## Session information

For reproducibility.

```{r,session}

sessionInfo()

```
