---
title: "GSE158433: GMEA of DNA methylation and RNA-seq differences in lung squamous cell carcinoma"
author: "The GMEA team"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_float: true
    code_folding: hide
    fig_width: 7
    fig_height: 7
theme: cosmo
---

## Introduction

The purpose of this analysis is to apply different enrichment analysis approaches,
and see which approach is most parsimonious with gene expressoin differences.

We will specifically be testing the GMEA approach versus existing approaches,
namely GOseq and mitch.

This study (GSE158433) is a super-series contains matched RNA-seq (GSE158420) and 
Infinium EPIC array (GSE158422) data for matched control and cancer tissues.

In this script, we will be processing the EPIC DNA methylation data and performing the
statistical analysis in limma and saving the results for downstream enrichment analysis.

```{r,packages}

suppressPackageStartupMessages({
  library("plyr")
  library("R.utils")
  library("missMethyl")
  library("limma")
  library("DMRcate")
  library("DMRcatedata")
  library("topconfects")
  library("minfi")
  library("IlluminaHumanMethylation450kmanifest")
  library("RColorBrewer")
  library("IlluminaHumanMethylation450kanno.ilmn12.hg19")
  library("GEOquery")
  library("eulerr")
  library("plyr")
  library("gplots")
  library("reshape2")
  library("forestplot")
  library("beeswarm")
  library("RCircos")
  library("qqman")
  library("ENmix")
  library("tictoc")
  library("mitch")
  library("kableExtra")
})

source("../meth_functions.R")
CORES=detectCores()/2

```

## Obtaining array annotations

```{r,annotation}

anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
myann <- data.frame(anno[,c("UCSC_RefGene_Name","Regulatory_Feature_Group","Islands_Name","Relation_to_Island")])
promoters <- grep("Prom",myann$Regulatory_Feature_Group)

```

## Load array result

GSE158422 is the accession number for the array data.

```{r,arraydataload}

#GSE158422 <- readRDS("GSE158422.rds")
dm <- read.table("GSE158422_limma.tsv",header=TRUE,sep="\t")

head(dm,50) %>%
  kbl(caption = "Top significant probes with limma") %>%
  kable_paper("hover", full_width = F)

rownames(dm) <- dm[,1]
dm <- dm[,c("UCSC_RefGene_Name","t")]

```

## GMEA whole gene

```{r,gmea_wg}

hist(dm$t)
tic() ; res <- calc_sc(dm) ; toc() #32 cores 95.5
res2 <- res[[2]]

head(res2,20) %>%
  kbl(caption = "Top significant genes with GMEA") %>%
  kable_paper("hover", full_width = F)

sig <- subset(res2,`fdr(sc)`<0.05)

es <- sig[order(-abs(sig$median)),]

head(es,20) %>%
  kbl(caption = "Top effect size probes with GMEA") %>%
  kable_paper("hover", full_width = F)

gmea_volc(res2)
gmea_boxplot(res)
res2_wg <- res2

```

## Mitch whole gene

Methylation only.

```{r,mitch_wg}

genesets <- gmt_import("../ReactomePathways.gmt")

meth <- res2$sig * res2$median
meth <- as.data.frame(meth)
rownames(meth) <- rownames(res2)

mres <- mitch_calc(x=meth, genesets=genesets,priority="effect")

head(mres$enrichment_result,20) %>%
  kbl(caption = "Top differential pathways with GMEA+mitch") %>%
  kable_paper("hover", full_width = F)

```

Methylation and gene expression.

```{r,integrate_wg}

dge <- read.table("GSE158420_dge.tsv")
dgem <- mitch_import(dge,DEtype="deseq2")

int <- merge(meth,dgem,by=0)
rownames(int) <- int$Row.names
int[,1]=NULL
names(int) <- c("meth","rna")

plot(int,xlim=c(-100,100))
abline(h=0,lty=2,lwd=2,col="red")
abline(v=0,lty=2,lwd=2,col="red")

uu <- length(which(int$meth>0 & int$rna>0))
ud <- length(which(int$meth>0 & int$rna<0))
dd <- length(which(int$meth<0 & int$rna<0))
du <- length(which(int$meth<0 & int$rna>0))

uu
ud
dd
du

uu + dd
ud + du

int_res <- mitch_calc(int,genesets=genesets,priority="effect")

int_res_tbl <- int_res$enrichment_result

head(int_res_tbl, 20) %>%
  kbl(caption = "Integrative results with mitch") %>%
  kable_paper("hover", full_width = F)

plot(int_res_tbl$s.meth,int_res_tbl$s.rna,ylab="RNA",xlab="meth")

abline(h=0,lty=2,lwd=2,col="red")
abline(v=0,lty=2,lwd=2,col="red")

uu <- length(which(int_res_tbl$s.meth>0 & int_res_tbl$s.rna>0))
ud <- length(which(int_res_tbl$s.meth>0 & int_res_tbl$s.rna<0))
dd <- length(which(int_res_tbl$s.meth<0 & int_res_tbl$s.rna<0))
du <- length(which(int_res_tbl$s.meth<0 & int_res_tbl$s.rna>0))

uu
ud
dd
du

uu + dd
ud + du

mitch_report(int_res,outfile="int_wholegene.html",overwrite=TRUE)

```

## GMEA promoter

```{r,gmea_promoter}

dm <- read.table("GSE158422_limma.tsv",header=TRUE,sep="\t")
dm <- dm[grep("Promot",dm$Regulatory_Feature_Group),]
dm <- dm[,c("UCSC_RefGene_Name","t")]

hist(dm$t)
tic() ; res <- calc_sc(dm) ; toc() #32 cores 17.7
res2 <- res[[2]]

head(res2,20) %>%
  kbl(caption = "Top significant promoters with GMEA") %>%
  kable_paper("hover", full_width = F)

sig <- subset(res2,`fdr(sc)`<0.05)

es <- sig[order(-abs(sig$median)),]

head(es,20) %>%
  kbl(caption = "Top effect size probes with GMEA") %>%
  kable_paper("hover", full_width = F)

gmea_volc(res2)
gmea_boxplot(res)
res2_p <- res2

```

## Mitch promoter

Methylation only

```{r,mitchpromoter1}

genesets <- gmt_import("../ReactomePathways.gmt")

pmeth <- res2$sig * res2$median
pmeth <- as.data.frame(pmeth)
rownames(pmeth) <- rownames(res2)

mres <- mitch_calc(x=pmeth, genesets=genesets,priority="effect")

head(mres$enrichment_result,20) %>%
  kbl(caption = "Top differential pathways with GMEA+mitch") %>%
  kable_paper("hover", full_width = F)

```

Methylation and gene expression.

```{r,integrate_promoter}

#dge <- read.table("GSE158420_dge.tsv")
#dgem <- mitch_import(dge,DEtype="deseq2")

int <- merge(pmeth,dgem,by=0)
rownames(int) <- int$Row.names
int[,1]=NULL
names(int) <- c("meth","rna")

plot(int)
abline(h=0,lty=2,lwd=2,col="red")
abline(v=0,lty=2,lwd=2,col="red")

uu <- length(which(int$meth>0 & int$rna>0))
ud <- length(which(int$meth>0 & int$rna<0))
dd <- length(which(int$meth<0 & int$rna<0))
du <- length(which(int$meth<0 & int$rna>0))

uu
ud
dd
du

uu + dd
ud + du

int_res <- mitch_calc(int,genesets=genesets,priority="effect")

int_res_tbl <- int_res$enrichment_result

head(int_res_tbl,20) %>%
  kbl(caption = "Integrative results with mitch") %>%
  kable_paper("hover", full_width = F)

int_res_tbl <- int_res_tbl[order(-abs(int_res_tbl$s.dist)),]

head(int_res_tbl,20) %>%
  kbl(caption = "Integrative results with mitch") %>%
  kable_paper("hover", full_width = F)

plot(int_res_tbl$s.meth,int_res_tbl$s.rna,ylab="RNA",xlab="meth")

abline(h=0,lty=2,lwd=2,col="red")
abline(v=0,lty=2,lwd=2,col="red")

uu <- length(which(int_res_tbl$s.meth>0 & int_res_tbl$s.rna>0))
ud <- length(which(int_res_tbl$s.meth>0 & int_res_tbl$s.rna<0))
dd <- length(which(int_res_tbl$s.meth<0 & int_res_tbl$s.rna<0))
du <- length(which(int_res_tbl$s.meth<0 & int_res_tbl$s.rna>0))

uu
ud
dd
du

uu + dd
ud + du

mitch_report(int_res,outfile="int_promoter.html",overwrite=TRUE)

```

## Session Information

```{r,session}

sessionInfo()

```
